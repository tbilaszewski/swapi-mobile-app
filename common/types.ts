import { PeopleResponse, SpaceshipResponse } from "../repository/model";

export type RootStackParamList = {
  Root: undefined;
  NotFound: undefined;
};

export type BottomTabParamList = {
  People: undefined;
  Starships: undefined;
};

export type TabOneParamList = {
  PeopleScreen: undefined;
};

export type TabTwoParamList = {
  StarshipsScreen: undefined;
};

export type PeopleWithMass = Pick<PeopleResponse, "mass">;

export type StarshipsWithCrew = Pick<SpaceshipResponse, "crew">;

export type Nullable<T> = null | T;

export type Optional<T> = undefined | T;

export type CardData = PeopleResponse | SpaceshipResponse;

export const isPeopleResponse = (
  response: CardData
): response is PeopleResponse =>
  (response as PeopleResponse).mass !== undefined;

export const isSpaceshipResponse = (
  response: CardData
): response is SpaceshipResponse =>
  (response as SpaceshipResponse).crew !== undefined;
