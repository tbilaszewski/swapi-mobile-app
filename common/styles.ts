import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
  container: {
    padding: 20,
    margin: 10,
    flexGrow: 0,
  },
  winnerCard: {
    borderColor: "green",
    borderWidth: 5,
  },
});
