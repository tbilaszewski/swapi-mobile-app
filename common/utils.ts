import { PeopleWithMass, StarshipsWithCrew } from "./types";

export const parseValueToFloatNumber = (value: string): number => {
  const numberValue = Number.parseFloat(value);
  return Number.isNaN(numberValue) ? -1 : numberValue;
};

export const findMaxValueIndex = (arr: number[]): number => {
  const maxValue = Math.max(...arr);
  if (maxValue < 0) {
    return -1;
  }
  return arr.findIndex((item) => item === maxValue);
};

const findMaxNaturalValue = <T extends { [key: string]: string }>(
  array: T[],
  key: keyof T
): number => {
  const masses = array.map((item) => item[key]).map(parseValueToFloatNumber);
  return findMaxValueIndex(masses);
};

export const findHeaviest = (people: PeopleWithMass[]) =>
  findMaxNaturalValue<PeopleWithMass>(people, "mass");

export const findMaxCrewSize = (starships: StarshipsWithCrew[]) =>
  findMaxNaturalValue<StarshipsWithCrew>(starships, "crew");
