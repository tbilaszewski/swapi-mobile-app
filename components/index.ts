export * from "./PeopleCard/";
export * from "./SpaceshipCard/";
export * from "./CardNotFound";
export * from "./ScoreDisplay";
