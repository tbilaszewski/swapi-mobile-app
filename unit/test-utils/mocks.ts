import { PeopleResponse, SpaceshipResponse } from "../../repository/model";

export const personDataMock: PeopleResponse = {
  name: "Luke Skywalker",
  height: "172",
  mass: "77",
  hair_color: "blond",
  skin_color: "fair",
  eye_color: "blue",
  birth_year: "19BBY",
  gender: "male",
  homeworld: "https://swapi.dev/api/planets/1/",
  films: [
    "https://swapi.dev/api/films/2/",
    "https://swapi.dev/api/films/6/",
    "https://swapi.dev/api/films/3/",
    "https://swapi.dev/api/films/1/",
    "https://swapi.dev/api/films/7/",
  ],
  species: ["https://swapi.dev/api/species/1/"],
  vehicles: [
    "https://swapi.dev/api/vehicles/14/",
    "https://swapi.dev/api/vehicles/30/",
  ],
  starships: [
    "https://swapi.dev/api/starships/12/",
    "https://swapi.dev/api/starships/22/",
  ],
  created: "2014-12-09T13:50:51.644000Z",
  edited: "2014-12-20T21:17:56.891000Z",
  url: "https://swapi.dev/api/people/1/",
};

export const spaceshipDataMock: SpaceshipResponse = {
  name: "Death Star",
  model: "DS-1 Orbital Battle Station",
  manufacturer:
    "Imperial Department of Military Research, Sienar Fleet Systems",
  cost_in_credits: "1000000000000",
  length: "120000",
  max_atmosphering_speed: "n/a",
  crew: "342,953",
  passengers: "843,342",
  cargo_capacity: "1000000000000",
  consumables: "3 years",
  hyperdrive_rating: "4.0",
  MGLT: "10",
  starship_class: "Deep Space Mobile Battlestation",
  pilots: [],
  films: ["http://swapi.dev/api/films/1/"],
  created: "2014-12-10T16:36:50.509000Z",
  edited: "2014-12-20T21:26:24.783000Z",
  url: "http://swapi.dev/api/starships/9/",
};
