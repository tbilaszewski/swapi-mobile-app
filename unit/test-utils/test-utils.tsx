import { render } from "@testing-library/react-native";
import { Provider } from "react-redux";
import React from "react";
import { store } from "../../store";

const AllTheProviders: React.FC = ({ children }) => {
  return <Provider store={store}>{children}</Provider>;
};

const customRender = (el: React.ReactElement, options = {}) =>
  render(el, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react-native";

export { customRender as render };
