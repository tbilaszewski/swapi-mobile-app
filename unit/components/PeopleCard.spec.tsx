import React from "react";
import { styles } from "../../common/styles";
import { PeopleCard } from "../../components";
import { render, personDataMock } from "../test-utils";

describe("PeopleCard", () => {
  it("should match snapshot ", () => {
    const cardWrapper = render(<PeopleCard data={personDataMock} />);
    expect(cardWrapper.toJSON()).toMatchSnapshot();
  });

  it("should have name rendered", () => {
    const { getByText } = render(<PeopleCard data={personDataMock} />);

    const name = getByText(personDataMock.name);
    expect(name).toBeTruthy();
  });

  it("should have crew rendered", () => {
    const { getByText } = render(<PeopleCard data={personDataMock} />);

    const name = getByText(`Mass: ${personDataMock.mass}`);
    expect(name).toBeTruthy();
  });

  it("should render card not found message when data not passed ", () => {
    const { getByText } = render(<PeopleCard />);

    expect(getByText("Cards not found")).toBeTruthy();
  });
});
