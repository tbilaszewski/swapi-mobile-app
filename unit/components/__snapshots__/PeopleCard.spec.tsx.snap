// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeopleCard should match snapshot  1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "borderRadius": 4,
      "elevation": 1,
      "flexGrow": 0,
      "margin": 10,
      "padding": 20,
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 0.5,
        "width": 0,
      },
      "shadowOpacity": 0.24,
      "shadowRadius": 0.75,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingLeft": 16,
          },
          Object {
            "minHeight": 50,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Array [
                    Object {
                      "minHeight": 30,
                    },
                    Object {
                      "marginBottom": 2,
                    },
                    undefined,
                  ],
                ],
              ],
            ]
          }
        >
          Luke Skywalker
        </Text>
      </View>
      <View />
    </View>
    <View
      style={
        Array [
          Object {
            "paddingHorizontal": 16,
          },
          Object {
            "paddingBottom": 16,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 14,
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                  "marginVertical": 2,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        Birth: 
        19BBY
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 14,
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                  "marginVertical": 2,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        Gender: 
        male
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 14,
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                  "marginVertical": 2,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        Height: 
        172
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 14,
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                  "marginVertical": 2,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        Mass: 
        77
      </Text>
    </View>
  </View>
</View>
`;
